libraries/AP_Scripting/drivers/EFI_NMEA2k.lua

```diff
@@ -0,0 +1,196 @@
+
+local PGN_TABLE = {
+   [PGN_ENGINE_PARAM_RAPID] = 8,
+   [PGN_ENGINE_PARAM_DYNAMIC] = 26
+}
+
+NMEA_2000.set_PGN_table(PGN_TABLE)
+
+local frame_count = 0
+local state = {}
+
+local function log_frame(frame)
+   local id = frame:id()
+   logger:write("CANF",'Id,DLC,FC,B0,B1,B2,B3,B4,B5,B6,B7','IBIBBBBBBBB',
+                id,
+                frame:dlc(),
+                frame_count,
+                frame:data(0), frame:data(1), frame:data(2), frame:data(3),
+                frame:data(4), frame:data(5), frame:data(6), frame:data(7))
+   frame_count = frame_count + 1
+end
+
+--[[
+   parse the higher rate engine data, giving RPM and pressure
+--]]
+local function parse_engine_param_rapid(data)
+   state.instance, state.speed, state.boost_presssure, state.tilt, _ = string.unpack("<BHHbH", data)
+   if not state.instance then
+      return
+   end
+
+   efi_state:engine_speed_rpm(uint32_t(state.speed/4))
+   efi_state:atmospheric_pressure_kpa(state.boost_presssure*0.001)
+   efi_state:cylinder_status(cylinder_state)
+   efi_state:last_updated_ms(millis())
+
+   -- Set the EFI_State into the EFI scripting driver
+   if efi_backend then
+      efi_backend:handle_scripting(efi_state)
+   end
+end
+
+--[[
+   parse the lower rate engine data, giving other engine data
+--]]
+local function parse_engine_param_dynamic(data)
+   state.instance, state.oil_press, state.oil_temp, state.temp_K, state.alt_V, state.fuel_rate, state.engine_s, state.cool_press,
+      state.fuel_press, state.res1, state.dstat1, state.dstat2, state.eload, state.etorque = string.unpack("<BHHHhhIHHBHHbb", data)
+   if not state.instance then
+      return
+   end
+   efi_state:intake_manifold_temperature(state.temp_K*0.01)
+   efi_state:oil_pressure(state.oil_press*0.01*0.001)
+   efi_state:oil_temperature(state.oil_temp*0.1)
+   cylinder_state:cylinder_head_temperature(state.temp_K*0.01)
+   efi_state:ignition_voltage(state.alt_V*0.01)
+   efi_state:fuel_consumption_rate_cm3pm(state.fuel_rate*0.1*1000.0/3600.0)
+   efi_state:intake_manifold_pressure_kpa(state.cool_press*100*0.001)
+   efi_state:fuel_pressure(state.fuel_press)
+   efi_state:engine_load_percent(state.eload)
+end
+
+function update()
+   if EFI_2K_ENABLE:get() < 1 then
+      return update, 500
+   end
+   if not efi_backend then
+      efi_backend = efi:get_backend(0)
+   end
+   if not efi_backend then
+      return update, 500
+   end
+
+   -- read up to 25 frames per update() call
+   for _ = 1, 25 do
+      local frame = can_driver:read_frame()
+      if not frame then
+         break
+      end
+      if option_enabled(OPTION_LOGGING) then
+         log_frame(frame)
+      end
+      pgn, data = NMEA_2000.parse(frame)
+      if pgn then
+         if pgn == PGN_ENGINE_PARAM_RAPID then
+            parse_engine_param_rapid(data)
+         elseif pgn == PGN_ENGINE_PARAM_DYNAMIC then
+            parse_engine_param_dynamic(data)
+         end
+      end
+   end
+
+   return update, 2
+end
+
+return update()
```

Comment:
Is it worth using the protected wrapper instead?

Reply:
Not really related to this PR. But I'm not a fan of just putting in the protected wrapper everywhere. With adequate testing there should be no need. We typically also just try the same thing again rather than doing any sort of reset to try and clear the error.